Allow saving to nexus for GenerateGroupingPowder

rboston628 opened the pull request:
**Description of work**

Edited `GenerateGroupingPowder` so that can save to binary files using `SaveNexusProcessed`, using `GroupingWorkspaces` instead of vectors-of-vectors.  The functionality to save to XML and PAR were removed to methods.

The groups are by default labeled according to the sector of the ring they fall in.  Additional labeling options are by spherical sector, within ring but in order without missing labels, or in ring in order without missing labels and distinguishing left/right sides.

**Purpose of work**
<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
-->

For [EWM 1371](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=1371), 

**Report to:** @mguthriem

**Summary of work**
<!-- Please provide a short, high level description of the work that was done.
-->

Make `GenerateGroupingPowder` able to save as nexus files.

Introduce new labeling methods for the groups.

**Further detail of work**
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

New property `"FileFormat"` which defaults to `"xml"`, with allowed values `"xml"`, `"nxs"`, or `"nx5"`.
Groups are assigned through a `GroupingWorkspace`, rather than through a vector-of-vectors as before.
Saving to XML, PAR, and Nexus are handled by helper methods, called at end of execution.
Saving to XML retains original method.
Saving to PAR retains original method.
Saving to Nexus use `SaveNexusProcessed` with the input grouping workspace.

The group labels can be assigned in one of four ways:
 - by dividing the polar angle into sectors, and labeling each pixel by its circular sector, treating opposite sides with the same number
 - by dividing the entire sphere into sectors, and labeling each pixel by its spherical sector
 - by order of increasing polar angle, treating opposite sides with same angle as same group
 - by order of increasing polar angle, treating opposite sides as distinct groups
 These are handled through the parameters AngleStep, AzimuthalStep, AzimuthalStart.

**Note**
In order to make consistent with the old convention used by `GenerateGroupingPowder`, and also be consistent with the contradicting convention in `GroupingWorkspace`, it was necessary to alter labeling of the legacy labler (CircularSectorLabelor) to avoid the label of 0 on small-angle groups.  This is because [groups labeled 0 are ignored](https://github.com/mantidproject/mantid/blob/81974dcb524cbc42f6009ccfe524818a84a3033d/Framework/DataObjects/src/GroupingWorkspace.cpp#LL40C1-L40C1) by `GroupingWorkspace`.  To be as consistent as possible with how the code already exists, the GroupingWorkspace labels start at 1, and the XML file labels start at 0.  This means for the same calculation using the legacy labler,  the group IDs inside the GroupingWorkspace and the XML file will be off by 1 from each other.

**To test:**

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

See testing function for CTest, creates the grouping powder workspace, saves as nexus, then loads from nexus and compares.

Tests of the labeling create instances of SNAPlite arrangements and check for correct labeling.

*There is no associated issue.*

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

rboston628 committed:
Allow saving to nexus for GenerateGroupingPowder

peterfpeterson committed:
Drop default value for output workspace name

peterfpeterson committed:
Extract the mapping of output file to par file into static function

peterfpeterson committed:
Rename variable and use std function for rounding

rboston628 committed:
labelor methods

peterfpeterson commented on Framework/DataHandling/src/GenerateGroupingPowder.cpp:
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
   positiveDouble->setLower(0.0);
   declareProperty("AngleStep", -1.0, positiveDouble, "The angle step for grouping, in degrees.");
-  declareProperty(std::make_unique<FileProperty>("GroupingFilename", "", FileProperty::Save, ".xml"),
+  declareProperty("AzimuthalStep", -1.0, positiveDouble, "The azimuthal angle step for grouping, in degrees.");
```
The default should be 360 (basically ignore everything) and the range should be (0, 360] (exclusive, inclusive]

peterfpeterson commented on Framework/DataHandling/src/GenerateGroupingPowder.cpp:
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
   declareProperty("AngleStep", -1.0, positiveDouble, "The angle step for grouping, in degrees.");
-  declareProperty(std::make_unique<FileProperty>("GroupingFilename", "", FileProperty::Save, ".xml"),
+  declareProperty("AzimuthalStep", -1.0, positiveDouble, "The azimuthal angle step for grouping, in degrees.");
+  declareProperty("AzimuthalStart", -1.0, positiveDouble, "The aimuthal angle start location, in degrees.");
```
The default should be `EMPTY_DBL()` or `0` (I don't know which is better), and the range should be (-360, 360) (exclusive, exclusive)

peterfpeterson commented on Framework/DataHandling/src/GenerateGroupingPowder.cpp:
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
+                  "The grouping workspace created");
+
+  declareProperty("NumberByAngle", true,
+                  "If true, divide sphere into groups by angle and step, number according to band.");
```
A second sentence would help explain what this does: "empty parts of the instrument will effectively have group numbers that do not exist in the output"

rboston628 committed:
changes to bounds and default values for angle properties

rboston628 committed:
pass to labelor  by const& for better performance

rboston628 committed:
add validator to prevent azimuthal step if PAR file generated

rboston628 committed:
edit to validator of PAR and azimuthal step

rboston628 committed:
add validateInputs to header

peterfpeterson committed:
Add snap lite tests to preserve old functionality

peterfpeterson committed:
Re-enable test

peterfpeterson committed:
Add more tests as agreed upon

Some of the expected detector center values (especially out of plane) may be incorrect.
However, this commit annotates those so one can tell that there is lower confidence.

rboston628 committed:
additional labelors for spherical sectors and split-side rings

rboston628 committed:
release notes for PR 35462

rboston628 committed:
remove TODOs for grouping powder test

rboston628 committed:
changes to grouping powder to appease cppcheck

rboston628 committed:
more changes to grouping powder to appease cppcheck

rboston628 committed:
change to force checks to run

rboston628 committed:
generate grouping powder, changes to cppcheck and names

rboston628 committed:
fixes to SofTwoThetaTOFTest to pass

quantumsteve commented on Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:
```diff
@@ -1308,4 +1309,56 @@ Mantid::DataObjects::Workspace2D_sptr create2DWorkspace123WithMaskedBin(int numH
+  auto xAxis = wsIn->getAxis(0);
+  xAxis->unit() = Mantid::Kernel::UnitFactory::Instance().create("dSpacing");
+
+  wsIn = std::dynamic_pointer_cast<MatrixWorkspace>(AnalysisDataService::Instance().retrieve(wkspName));
```
How do lines 1338 and 1360 differ? Can one delete this line?

rboston628 replied:
I didn't write this part, but my understanding is that line 1338 uses `wsIn` to set properties for an input before an algorithm, while line 1360 is retrieving the output after the instrument has been loaded.

quantumsteve replied:
It's pointer semantics, so what you're retrieving should point to the same object.

rboston628 replied:
It should, but does it actually?

rboston628 replied:
comparing them with `ws->id()` indicates they are the same, and no checks failed when this line is removed

quantumsteve commented on Framework/DataHandling/src/GenerateGroupingPowder.cpp:
```diff
@@ -70,52 +278,94 @@ void GenerateGroupingPowder::exec() {
-  std::vector<std::vector<detid_t>> groups(numSteps);
-  std::vector<double> twoThetaAverage(numSteps, 0.);
-  std::vector<double> rAverage(numSteps, 0.);
+  Labelor *label;
```
```suggestion
  std::unique_ptr<Labelor> label;
```
There's a memory leak if line 331 throws.

rboston628 replied:
fixed

rboston628 committed:
changes to account for possible groupID=0 in old labelor

rboston628 committed:
fixedtypo

rboston628 committed:
suppress cppcheck complaining about unchanged 9-year-old code

rboston628 committed:
suppress cppcheck complaining about more unchanged 9-year-old code

rboston628 committed:
trying to make cpp check stop

rboston628 committed:
cppcheck again

rboston628 committed:
make cppcheck stop

rboston628 committed:
update cppcheck suppression error

rboston628 committed:
updated documentation for changes to GenerateGroupingPowder

rboston628 committed:
update to doc cleanup example, remove grouping workspace
